File: basic-web-validations/src/main/java/com/example/basicwebvalidations/by_annotation/UserController.java
package com.example.basicwebvalidations.by_annotation;

import org.springframework.validation.BeanPropertyBindingResult;
import org.springframework.validation.BindingResult;
import org.springframework.validation.ObjectError;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;
import java.util.List;

@RestController("byAnnotationController")
@RequestMapping("/users")
public class UserController {

    @PostMapping("/annotation")
    public List<ObjectError> byAnnotation(@Valid @RequestBody User user, BeanPropertyBindingResult errors){

        return errors.getAllErrors();
    }

}

File: basic-validations/src/main/kotlin/com/example/basicvalidations/UserController.kt
package com.example.basicvalidations

import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController
import javax.validation.Valid

@RestController
class UserController {

    @PostMapping("/users/annotation")
    fun byAnnotation(@Valid @RequestBody user: com.example.basicvalidations.by_annotation.User){}

}